File: LLMgeneratedCode/o3-high-1.R
# Install and load required packages
#if (!requireNamespace("BiocManager", quietly = TRUE))
#  install.packages("BiocManager")

#if (!require("GEOquery")) BiocManager::install("GEOquery")
#if (!require("glmnet")) install.packages("glmnet")

library(GEOquery)
library(glmnet)

# Download the GEO dataset (GSE149440)
# Setting GSEMatrix=TRUE returns an ExpressionSet object.
gset <- getGEO("GSE149440", GSEMatrix = TRUE)
# If multiple ExpressionSets are returned, select the first one.
if (length(gset) > 1) {
  eset <- gset[[1]]
} else {
  eset <- gset[[1]]
}

# Extract expression data and phenotype (sample metadata)
expr_data <- exprs(eset)       # rows: genes, columns: samples
pheno_data <- pData(eset)

# Check that the expected metadata columns exist
if (!all(c("gestational age:ch1", "train:ch1") %in% colnames(pheno_data))) {
  stop("The required metadata columns 'gestational age:ch1' and 'train:ch1' are not found.")
}

# Subset training and test samples based on the 'train:ch1' column
train_idx <- which(pheno_data$`train:ch1` == "1")
test_idx  <- which(pheno_data$`train:ch1` == "0")

# Prepare predictor matrices: glmnet expects rows=samples and columns=features
X_train <- t(expr_data[, train_idx])
X_test  <- t(expr_data[, test_idx])

# Prepare response vectors (convert to numeric in case they are factors/characters)
y_train <- as.numeric(as.character(pheno_data[train_idx, "gestational age:ch1"]))
y_test  <- as.numeric(as.character(pheno_data[test_idx, "gestational age:ch1"]))

# Fit a LASSO regression model (alpha = 1) using cross-validation to select lambda
set.seed(123)  # for reproducibility
cvfit <- cv.glmnet(X_train, y_train, alpha = 1)
best_lambda <- cvfit$lambda.min

# Train the final model using the best lambda
model <- glmnet(X_train, y_train, alpha = 1, lambda = best_lambda)

# Apply the model to the test set to obtain predictions
predictions <- predict(model, newx = X_test)

# Compute the Root Mean Squared Error (RMSE) on the test set
rmse <- sqrt(mean((predictions - y_test)^2))
cat("Test RMSE:", round(rmse, 2), "\n")
write.csv(data.frame(y_test=y_test,y_pred=predictions),file="./out2/o3-high-1.R.csv")

# Generate a scatter plot of predicted vs. actual gestational age values for the test set
plot(y_test, predictions,
     xlab = "Actual Gestational Age",
     ylab = "Predicted Gestational Age",
     main = paste("Predicted vs Actual Gestational Age (Test Set)\nRMSE =", round(rmse, 2)),
     pch = 19, col = "blue")
abline(0, 1, col = "red", lwd = 2)  # reference line (ideal prediction)

File: LLMgeneratedCode/o3-high-1.py
import GEOparse
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_squared_error

def main():
    # ---------------------------
    # 1. Download the GEO dataset
    # ---------------------------
    print("Downloading GEO dataset GSE149440 ...")
    gse = GEOparse.get_GEO("GSE149440", destdir="./")

    # ------------------------------------------------
    # 2. Extract the expression data (gene expression)
    # ------------------------------------------------
    # GEOparse can pivot the data so that rows=genes and columns=samples.
    # We then transpose to have samples as rows (each with tens of thousands of gene features).
    expr_df = gse.pivot_and_annotate('VALUE')
    X = expr_df.T

    # Optionally, if there are any non-numeric annotation columns in X, keep only numeric ones.
    X = X.select_dtypes(include=[np.number])

    # ----------------------------------------
    # 3. Build a metadata DataFrame from GSMs
    # ----------------------------------------
    # For each sample (GSM), extract the metadata entries:
    #   - "train:ch1" (used to split training/test)
    #   - "gestational age:ch1" (target variable)
    metadata = {}
    for gsm_name, gsm in gse.gsms.items():
        md = gsm.metadata
        train_val = md.get("train:ch1", [None])[0]
        ga_val = md.get("gestational age:ch1", [None])[0]
        metadata[gsm_name] = {"train": train_val, "gestational_age": ga_val}
    metadata_df = pd.DataFrame.from_dict(metadata, orient='index')

    # Ensure the samples in the expression matrix and metadata match
    common_samples = X.index.intersection(metadata_df.index)
    X = X.loc[common_samples]
    metadata_df = metadata_df.loc[common_samples]

    # Convert gestational age to numeric (if it isn’t already)
    metadata_df['gestational_age'] = pd.to_numeric(metadata_df['gestational_age'], errors='coerce')

    # Drop any samples with missing gestational age or train flag
    metadata_df = metadata_df.dropna(subset=['gestational_age', 'train'])
    X = X.loc[metadata_df.index]

    # ------------------------------------------------
    # 4. Split the data into training and test sets
    # ------------------------------------------------
    # Training set: samples where train:ch1 == '1'
    # Test set: samples where train:ch1 == '0'
    train_samples = metadata_df[metadata_df['train'] == '1'].index
    test_samples  = metadata_df[metadata_df['train'] == '0'].index

    X_train = X.loc[train_samples].astype(float)
    y_train = metadata_df.loc[train_samples, 'gestational_age']

    X_test = X.loc[test_samples].astype(float)
    y_test = metadata_df.loc[test_samples, 'gestational_age']

    # ------------------------------------------------
    # 5. Train a prediction model (Ridge Regression)
    # ------------------------------------------------
    model = Ridge(alpha=1.0)
    print("Fitting the regression model on the training set ...")
    model.fit(X_train, y_train)

    # Predict gestational age on the test set
    y_pred = model.predict(X_test)

    # --------------------------------------
    # 6. Evaluate the model (compute RMSE)
    # --------------------------------------
    rmse = np.sqrt(mean_squared_error(y_test, y_pred))
    print("Test RMSE: {:.2f}".format(rmse))

    # --------------------------------------------------------
    # 7. Generate a scatter plot: Predicted vs Actual values
    # --------------------------------------------------------
    plt.figure(figsize=(8, 6))
    plt.scatter(y_test, y_pred, alpha=0.7, edgecolor='k')
    plt.xlabel("Actual Gestational Age")
    plt.ylabel("Predicted Gestational Age")
    plt.title("Predicted vs Actual Gestational Age (Test Set)")
    # Plot the diagonal line for reference
    min_val = min(y_test.min(), y_pred.min())
    max_val = max(y_test.max(), y_pred.max())
    plt.plot([min_val, max_val], [min_val, max_val], 'r--', lw=2)
    plt.tight_layout()
    plt.savefig("./out2/o3-high-1.py.pdf", format="pdf")

if __name__ == "__main__":
    main()
